File: labsoundpy/labsound/nodes/destination_node.py
"""
AudioDestinationNode class for LabSound Python bindings.
"""

from typing import Optional, Union

from ..audio_node import AudioNode


class AudioDestinationNode(AudioNode):
    """
    The AudioDestinationNode interface represents the end destination of an audio
    graph in a given context — usually the speakers of your device.
    
    This is the final node in the audio graph, and it cannot be connected to other nodes.
    """
    
    def __init__(self, context, node):
        """
        Initialize a new AudioDestinationNode.
        
        Args:
            context: The AudioContext this node belongs to
            node: The underlying C++ node object
        """
        super().__init__(context, node)
    
    @property
    def max_channel_count(self) -> int:
        """
        Get the maximum number of channels that this destination can support.
        
        Returns:
            The maximum number of channels
        """
        return self._node.max_channel_count()
    
    def set_channel_count(self, count: int):
        """
        Set the number of channels for this destination.
        
        Args:
            count: The number of channels
        """
        self._node.set_channel_count(count)

File: labsoundpy/src/nodes/destination_node.cpp
/**
 * Implementation file for AudioDestinationNode bindings.
 */

#include "nodes/destination_node.h"
#include <nanobind/stl/string.h>
#include <nanobind/stl/vector.h>
#include <nanobind/stl/shared_ptr.h>

void register_destination_node(nb::module_ &m) {
    // Bind the AudioDestinationNode class
    nb::class_<lab::AudioDestinationNode, lab::AudioNode, std::shared_ptr<lab::AudioDestinationNode>>(m, "_AudioDestinationNode")
        .def("max_channel_count", &lab::AudioDestinationNode::maxChannelCount)
        .def("set_channel_count", &lab::AudioDestinationNode::setChannelCount, nb::arg("count"));
}
